libs/platform/ebpf_epoch.c

```diff
@@ -61,25 +70,31 @@ typedef struct _ebpf_epoch_state
     bool timer_disabled : 1; // Prevent re-arming the timer during shutdown.
 } ebpf_epoch_state_t;
 
+typedef struct _ebpf_epoch_thread_entry
+{
+    ebpf_epoch_state_t epoch_state;     // Per-thread epoch state.
+    uintptr_t old_thread_affinity_mask; // Thread affinity mask before entering an entry/exit block.
+    uintptr_t thread_id;                // Thread ID of the thread that owns this entry or 0 if unused.
+} ebpf_epoch_thread_entry_t;
+
+// The epoch code limits the number of threads that can be active at any one time per CPU.
+// If additional threads attempt to call ebpf_epoch_enter, they will block until a thread exits the epoch.
+#define EBPF_EPOCH_THREAD_ENTRY_TABLE_SIZE ((EBPF_CACHE_LINE_SIZE * 3) / sizeof(ebpf_epoch_thread_entry_t))
+
 // Table to track per CPU state.
 // This table must fit into a multiple of EBPF_CACHE_LINE_SIZE.
-typedef struct _ebpf_epoch_cpu_entry
+#pragma warning(disable : 4324) // Structure was padded due to alignment specifier.
+typedef __declspec(align(EBPF_CACHE_LINE_SIZE)) struct _ebpf_epoch_cpu_entry
 {
     ebpf_lock_t lock;
     _Guarded_by_(lock) ebpf_epoch_state_t epoch_state;                 // Per-CPU epoch state.
     _Guarded_by_(lock) ebpf_list_entry_t free_list;                    // Per-CPU free list.
-    _Guarded_by_(lock) ebpf_hash_table_t* thread_table;                // Per-CPU thread table.
     _Guarded_by_(lock) ebpf_non_preemptible_work_item_t* stale_worker; // Per-CPU stale worker DPC.
-    uint32_t padding;                                                  // Pad to multiple of EBPF_CACHE_LINE_SIZE.
+    ebpf_semaphore_t* thread_entry_table_semaphore; // Semaphore to number of threads active in the epoch per CPU.
+    _Guarded_by_(lock)
+        ebpf_epoch_thread_entry_t thread_entry_table[EBPF_EPOCH_THREAD_ENTRY_TABLE_SIZE]; // Per-thread epoch state.
```

Comment:
Would it help if we were to ensure that each thread entry in this table were to be cache line aligned as well?  Given the high frequency access of these tables, would it be worthwhile ensuring optimal cache-line and/or native size alignments, including resizing (and where possible even re-ordering of individual members) of the structures involved?

Reply:
Resolved per discussion with Alan.